File: swift/hanoi_tower.swift
import Foundation

var moveCount = 0

func hanoiTower(_ n: Int, _ from: Character, _ to: Character, _ aux: Character) {
    if n == 1 {
        moveCount += 1
        return
    }
    
    hanoiTower(n - 1, from, aux, to)
    moveCount += 1
    hanoiTower(n - 1, aux, to, from)
}

let args = CommandLine.arguments
if args.count < 2 {
    print("Usage: hanoi_tower <n>")
    exit(1)
}

guard let n = Int(args[1]) else {
    print("Invalid number: \(args[1])")
    exit(1)
}

moveCount = 0

let start = CFAbsoluteTimeGetCurrent()
hanoiTower(n, "A", "C", "B")
let end = CFAbsoluteTimeGetCurrent()

let timeMs = Int((end - start) * 1000)

print("Swift: hanoi_tower(\(n)) = \(moveCount)")
print("Time: \(timeMs)ms")

File: lua/hanoi_tower.lua
local move_count = 0

function hanoi_tower(n, from, to, aux)
    if n == 1 then
        move_count = move_count + 1
        return
    end
    
    hanoi_tower(n - 1, from, aux, to)
    move_count = move_count + 1
    hanoi_tower(n - 1, aux, to, from)
end

if #arg < 1 then
    print('Usage: hanoi_tower <n>')
    os.exit(1)
end

local n = tonumber(arg[1])
if not n then
    print('Invalid number: ' .. arg[1])
    os.exit(1)
end

move_count = 0

local start = os.clock()
hanoi_tower(n, 'A', 'C', 'B')
local finish = os.clock()

local time_ms = math.floor((finish - start) * 1000)

print('Lua: hanoi_tower(' .. n .. ') = ' .. move_count)
print('Time: ' .. time_ms .. 'ms')